src/access/AccessRestrictionProvider.h

```diff
@@ -0,0 +1,283 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "RequestPath.h"
+#include "SubjectDescriptor.h"
+#include "Privilege.h"
+#include <algorithm>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <cstdint>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CHIPMem.h>
+#include <map>
+#include <memory>
+#include <protocols/interaction_model/Constants.h>
+#include <vector>
+
+namespace chip {
+namespace Access {
+
+class AccessRestrictionProvider
+{
+public:
+    static constexpr size_t kNumberOfFabrics      = CHIP_CONFIG_MAX_FABRICS;
+    static constexpr size_t kEntriesPerFabric     = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC;
+    static constexpr size_t kRestrictionsPerEntry = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY;
+
+    /**
+     * Defines the type of access restriction, which is used to determine the meaning of the restriction's id.
+     */
+    enum class Type : uint8_t
+    {
+        kAttributeAccessForbidden = 0,
+        kAttributeWriteForbidden  = 1,
+        kCommandForbidden         = 2,
+        kEventForbidden           = 3
+    };
+
+    /**
+     * Defines a single restriction on an attribute, command, or event.
+     *
+     * If id is not set, the restriction applies to all attributes, commands, or events of the given type (wildcard).
+     */
+    struct Restriction
+    {
+        Type restrictionType;
+        Optional<uint32_t> id;
+    };
+
+    /**
+     * Defines a single entry in the access restriction list, which contains a list of restrictions
+     * for a cluster on an endpoint.
+     */
+    struct Entry
+    {
+        FabricIndex fabricIndex;
+        EndpointId endpointNumber;
+        ClusterId clusterId;
+        std::vector<Restriction> restrictions;
+    };
+
+    /**
+     * Defines the interface for a checker for access restriction exceptions.
+     */
+    class AccessRestrictionExceptionChecker
+    {
+        public:
+            virtual ~AccessRestrictionExceptionChecker() = default;
+
+            /**
+             * Check if any restrictions are allowed to be applied to the given request.
+             *
+             * @retval true if restrictions may NOT be applied
+             */
```

Comment:
```suggestion
             * Check if any restrictions are disallowed to be applied to the given request
             * because of constraints against their use in ARLs.
             *
             * @retval true if ARL checks should be skipped for the path, false otherwise
             */
```

Reply:
done